baselines/a2c/utils.py

```diff
@@ -152,131 +41,44 @@ def discount_with_dones(rewards, dones, gamma):
-            return np.mean(self.lenbuffer)
-        else:
-            return 0  # on the first params dump, no episodes are finished
-
-    def mean_reward(self):
-        if self.rewbuffer:
-            return np.mean(self.rewbuffer)
-        else:
-            return 0
-
-
-# For ACER
-def get_by_index(x, idx):
-    assert(len(x.get_shape()) == 2)
-    assert(len(idx.get_shape()) == 1)
-    idx_flattened = tf.range(0, x.shape[0]) * x.shape[1] + idx
-    y = tf.gather(tf.reshape(x, [-1]),  # flatten input
-                  idx_flattened)  # use flattened indices
-    return y
-
-def check_shape(ts,shapes):
-    i = 0
-    for (t,shape) in zip(ts,shapes):
-        assert t.get_shape().as_list()==shape, "id " + str(i) + " shape " + str(t.get_shape()) + str(shape)
-        i += 1
-
-def avg_norm(t):
-    return tf.reduce_mean(tf.sqrt(tf.reduce_sum(tf.square(t), axis=-1)))
-
-def gradient_add(g1, g2, param):
-    print([g1, g2, param.name])
-    assert (not (g1 is None and g2 is None)), param.name
-    if g1 is None:
-        return g2
-    elif g2 is None:
-        return g1
-    else:
-        return g1 + g2
-
-def q_explained_variance(qpred, q):
-    _, vary = tf.nn.moments(q, axes=[0, 1])
-    _, varpred = tf.nn.moments(q - qpred, axes=[0, 1])
-    check_shape([vary, varpred], [[]] * 2)
-    return 1.0 - (varpred / vary)
+class InverseLinearTimeDecay(tf.keras.optimizers.schedules.LearningRateSchedule):
+    def __init__(self, initial_learning_rate, nupdates, name="InverseLinearTimeDecay"):
+        super(InverseLinearTimeDecay, self).__init__()
+        self.initial_learning_rate = initial_learning_rate
+        self.nupdates = nupdates
+        self.name = name
+
+    def __call__(self, step):
+        with tf.name_scope(self.name):
+            initial_learning_rate = tf.convert_to_tensor(self.initial_learning_rate, name="initial_learning_rate")
+            dtype = initial_learning_rate.dtype
+            step_t = tf.cast(step, dtype)
+            nupdates_t = tf.convert_to_tensor(self.nupdates, dtype=dtype)
+            tf.assert_less(step_t, nupdates_t)
+            return initial_learning_rate * (1. - step_t / nupdates_t)
+
+    def get_config(self):
+        return {
+            "initial_learning_rate": self.initial_learning_rate,
+            "nupdates": self.nupdates,
+            "name": self.name
+        }
+
+class LinearTimeDecay(tf.keras.optimizers.schedules.LearningRateSchedule):
+    def __init__(self, initial_learning_rate, name="LinearTimeDecay"):
+        super(LinearTimeDecay, self).__init__()
+        self.initial_learning_rate = initial_learning_rate
+        self.name = name
+
+    def __call__(self, step):
+        with tf.name_scope(self.name):
+            initial_learning_rate = tf.convert_to_tensor(self.initial_learning_rate, name="initial_learning_rate")
+            dtype = initial_learning_rate.dtype
+            step_t = tf.cast(step, dtype)
+            return initial_learning_rate * step_t
```

Comment:
The name LinearTimeDecay seems a little confusing, given that the learning rate actually increases with time

Reply:
Yeah it is...I was using it for PPO, turns out we don't need this anymore. Removed.